src/go/rdctl/pkg/config/config.go

```diff
@@ -0,0 +1,117 @@
+/*
+Copyright © 2022 SUSE LLC
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// Package config handles all the config-related parts of rdctl
+
+package config
+
+import (
+	"encoding/json"
+	"fmt"
+	"github.com/spf13/cobra"
+	"io/ioutil"
+	"log"
+	"os"
+	"path/filepath"
+	"strconv"
+)
+
+// The CLIConfig struct is used to store the json data read from the config file.
+type CLIConfig struct {
+	User     string
+	Password string
+	Port     int
+}
+
+var (
+	user     string
+	host     string
+	port     string
+	password string
+
+	configDir           string
+	configPath          string
+	defaultConfigPath   string
+	deferredConfigError error
+)
+
+// DefineGlobalFlags sets up the global flags, available for all sub-commands
+func DefineGlobalFlags(rootCmd *cobra.Command) {
+	var err error
+
+	configDir, err = os.UserConfigDir()
+	if err != nil {
+		log.Fatal("Can't get config-dir: ", err)
+	}
+	defaultConfigPath = filepath.Join(configDir, "rancher-desktop", "rd-engine.json")
+	rootCmd.PersistentFlags().StringVar(&configPath, "config-path", "", fmt.Sprintf("config file (default %s)", defaultConfigPath))
+	rootCmd.PersistentFlags().StringVar(&user, "user", "", "overrides the user setting in the config file")
+	rootCmd.PersistentFlags().StringVar(&host, "host", "", "default is localhost; most useful for WSL")
+	rootCmd.PersistentFlags().StringVar(&port, "port", "", "overrides the port setting in the config file")
+	rootCmd.PersistentFlags().StringVar(&password, "password", "", "overrides the password setting in the config file")
+}
+
+// GetConnectionInfo returns the connection info if it has it, and an error message explaining why
+// it isn't available if it doesn't have it.
+func GetConnectionInfo() (string, string, string, string, error) {
+	if deferredConfigError != nil && insufficientConnectionInfo() {
+		return "", "", "", "", deferredConfigError
+	}
+	return host, port, user, password, nil
+}
+
+// InitConfig is run after all modules are loaded and before the appropriate Execute function is invoked
+func InitConfig() {
```

Comment:
This doesn't need to be exported, right? Just called from within `GetConnectionInfo` (possibly via a `sync.Once`)?

Reply:
Well, the function needs to be invoked after the flags have been processed, but before the subcommand's `RunE` method is invoked.  From the cobra code:
```
// OnInitialize sets the passed functions to be run when each command's
// Execute method is called.
```

This does exactly what we want. I could experiment with a different approach, but this isn't a typical Cobra-based approach, and I don't see what it accomplishes.

Reply:
But it doesn't need to run before `RunE`; it just needs to be done before `GetConnectionInfo` returns (because that's the only point at which other packages can get the results).

> This does exactly what we want. I could experiment with a different approach, but this isn't a typical Cobra-based approach, and I don't see what it accomplishes.

The typical cobra-based approach involves _zero knowledge_ of cobra outside `cmd/`. So this is already a non-standard approach; it would be better to excise cobra from here completely.

Reply:
Once again it turns out following Mark's suggestion makes for cleaner code. About to push a commit that gets rid of the `deferredConfigError` variable, should be more straightforward.